utils/zerovec/src/zerovec/slice.rs

```diff
@@ -0,0 +1,318 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use super::*;
+use alloc::boxed::Box;
+
+/// A zero-copy "slice", i.e. the zero-copy version of `[T]`. This behaves
+/// similarly to [`ZeroVec<T>`], however [`ZeroVec<T>`] is allowed to contain
+/// owned data and as such is ideal for deserialization since most human readable
+/// serialization formats cannot unconditionally deserialize zero-copy.
+///
+/// This type can be used inside `VarZeroVec<T>`: This essentially allows for
+/// the construction of zero-copy types isomorphic to `Vec<Vec<T>>` by instead
+/// using `VarZeroVec<ZeroSlice<T>>`. See the [`VarZeroVec`](crate::VarZeroVec) docs for an example.
+#[repr(transparent)]
+pub struct ZeroSlice<T: AsULE>([T::ULE]);
+
+impl<T> ZeroSlice<T>
+where
+    T: AsULE,
+{
+    /// Get this [`ZeroSlice`] as a borrowed [`ZeroVec`]
+    ///
+    /// [`ZeroSlice`] does not have most of the methods that [`ZeroVec`] does,
+    /// so it is recommended to convert it to a [`ZeroVec`] before doing anything.
+    #[inline]
+    pub fn as_zerovec(&self) -> ZeroVec<'_, T> {
+        ZeroVec::Borrowed(&self.0)
+    }
+
+    /// Construct a `&ZeroSlice<T>` from a slice of ULEs
+    #[inline]
+    pub fn from_slice(slice: &[T::ULE]) -> &Self {
+        // This is safe because ZeroSlice is transparent over [T::ULE]
+        // so &ZeroSlice<T> can be safely cast from &[T::ULE]
+        unsafe { &*(slice as *const _ as *const Self) }
+    }
+
+    /// Construct a `Box<ZeroSlice<T>>` from a boxed slice of ULEs
+    #[inline]
+    pub fn from_boxed_slice(slice: Box<[T::ULE]>) -> Box<Self> {
+        // This is safe because ZeroSlice is transparent over [T::ULE]
+        // so Box<ZeroSlice<T>> can be safely cast from Box<[T::ULE]>
+        unsafe { Box::from_raw(Box::into_raw(slice) as *mut Self) }
+    }
+
+    /// Returns this slice as its underlying `&[u8]` byte buffer representation.
+    ///
+    /// Useful for serialization.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use zerovec::ZeroVec;
+    ///
+    /// // The little-endian bytes correspond to the numbers on the following line.
+    /// let bytes: &[u8] = &[0xD3, 0x00, 0x19, 0x01, 0xA5, 0x01, 0xCD, 0x01];
+    /// let nums: &[u16] = &[211, 281, 421, 461];
+    ///
+    /// let zerovec = ZeroVec::alloc_from_slice(nums);
+    ///
+    /// assert_eq!(bytes, zerovec.as_bytes());
+    /// ```
+    #[inline]
+    pub fn as_bytes(&self) -> &[u8] {
+        T::ULE::as_byte_slice(self.as_slice())
+    }
+
+    /// Dereferences this slice as `&[T::ULE]`.
+    #[inline]
+    pub fn as_slice(&self) -> &[T::ULE] {
```

Comment:
Nit (related to above suggestion): This is not an API people should generally be calling any more.  We should not encourage people to access `&[T::ULE]`.  The `as_slice()` function on ZeroVec should therefore return a ZeroSlice.

Consider removing this function and making the enum field public?  Or rename it to `as_ule_slice()`